transport/src/main/java/io/netty/channel/DefaultChannelPromiseAggregatorFactory.java

```diff
@@ -0,0 +1,173 @@
+        super(aggregationResult.channel(), executor);
+        assert !aggregationResult.isDone();
+        this.promise = aggregationResult;
+    }
+
+    /**
+     * Allocate a new promise which will be used to aggregate the overall success of this promise aggregator.
+     * @return A new promise which will be aggregated.
+     * {@code null} if {@link #doneAllocatingPromises()} was previously called.
+     */
+    public ChannelPromise newPromise() {
+        assert !doneAllocating : "Done allocating. No more promises can be allocated.";
+        ++expectedCount;
+        return this;
+    }
+
+    /**
+     * Signify that no more {@link #newPromise()} allocations will be made.
+     * The aggregation can not be successful until this method is called.
+     * @return The promise that is the aggregation of all promises allocated with {@link #newPromise()}.
+     */
+    public ChannelPromise doneAllocatingPromises() {
+        if (!doneAllocating) {
+            doneAllocating = true;
+            if (doneCount == expectedCount || expectedCount == 0) {
+                return setPromise();
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public boolean tryFailure(Throwable cause) {
+        if (allowFailure()) {
+            ++doneCount;
+            lastFailure = cause;
+            if (allPromisesDone()) {
+                return tryPromise();
+            }
+            // TODO: We break the interface a bit here, but we can't return false because according to the javadoc
+            // returning false can mean "already marked as either a success or a failure".
+            // Multiple failure events can be processed without issue because this is an aggregation.
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Fail this object if it has not already been failed.
+     * <p>
+     * This method will NOT throw an {@link IllegalStateException} if called multiple times
+     * because that may be expected.
+     */
+    @Override
+    public ChannelPromise setFailure(Throwable cause) {
+        if (allowFailure()) {
+            ++doneCount;
+            lastFailure = cause;
+            if (allPromisesDone()) {
+                return setPromise();
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public ChannelPromise setSuccess(Void result) {
+        if (awaitingPromises()) {
+            ++doneCount;
+            if (allPromisesDone()) {
+                setPromise();
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public boolean trySuccess(Void result) {
+        if (awaitingPromises()) {
+            ++doneCount;
+            if (allPromisesDone()) {
+                return tryPromise();
+            }
+            // TODO: We break the interface a bit here, but we can't return false because according to the javadoc
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)